### PR TITLE
[WIP] Implement XBM

### LINKED ISSUE #163: Implement cross-batch memory for losses
- Paper: https://arxiv.org/pdf/1912.06798.pdf
- Reference for implementation: https://github.com/msight-tech/research-xbm/
## How it works
- XBM relies on the observation that the drift of embeddings is slow during training, i.e., embeddings for the same object is changing in a very slow pace.
- This lets us add embeddings and targets in a ring buffer of a certain size.
- After a certain number of iterations, start using the buffer. Now the final loss is the weighted sum of the actual mini-batch loss and the ring buffer loss.

## Suggested implementation
1. Introduce an `XBMConfig` class to hold the configuration values such as `buffer_size`, `start_iteration`, `xbm_weight`.
2. Add a `configure_xbm()` hook in `TrainableModel` and return `None` by default.
3. İf it returns an `XBMConfig` instance instead, create a `XBMBuffer` instance in the `TrainableModel` constructor.
4. Implement the XBM logic in `_common_step` if `stage` is training.

## Notes
1. We cannot re-use the existing `Accumulator` classes because they are not ring buffers.
2. I don't think we need a mixin because addition to `TrainableModel` will be only a few lines of code, and we need to update `_common_step` anyway.

### PR DESCRIPTION
**Pull request by monatis**
Closes #163.

I'm adding commits after small refactoring and polishing comments. Will request a review afterwards.

### CODE REVIEW

**Review by joein — CHANGES_REQUESTED** [review]
it seems like cifar examples deserve a separate folder already :)

it will also allow us to avoid code duplication, because the only differences I see are some imports, `Model` class and some additional cmd options

File: quaterion/train/xbm/xbm_buffer.py [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+    """A buffer implementation to hold recent N embeddings and target values.
+
+    Inspired by https://github.com/msight-tech/research-xbm/blob/master/ret_benchmark/modeling/xbm.py
+
+    Args:
+        config: Config class to configure XBM settings.
+    """
+
+    def __init__(self, config: XbmConfig):
+        if config.device == XbmDevice.AUTO:
```
[the remark this answers is not in the document]
Comment (monatis) [reply quaterion/train/xbm/xbm_buffer.py:19]: I'm afraid it's `CacheType` whose behavior should be changed because [it's not reliable to get device from a module](https://stackoverflow.com/questions/58926054/how-to-get-the-device-type-of-a-pytorch-module-conveniently).

Others are good catches --already fixes some of them, and I'll continue to fix others tomorrow.

File: quaterion/train/trainable_model.py [hunk 2]
```diff
@@ -196,7 +205,7 @@ def configure_caches(self) -> Optional[CacheConfig]:
 
         `Do not use cache (default)`::
 
-            return None
+           < return None
```
[the remark this answers is not in the document]
Comment (monatis) [reply quaterion/train/trainable_model.py]: apparently just an accident during fast scrolling.

File: quaterion/train/xbm/xbm_config.py [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+    """Determine XBM settings.
+
+    This class should be returned from
+    :meth:`~quaterion.train.trainable_model.TrainableModel.configure_xbm`
+    """
+
+    embedding_size: int
+    """Dimensionality of embeddings outputted by the encoder"""
+
+    weight: Optional[float] = 1.0
```
[the remark this answers is not in the document]
Comment (monatis) [reply quaterion/train/xbm/xbm_config.py:27]: It's the default value in their implementation. Most often, this will be combined with a scheduler to decay LR, but I chose to simplify the example with a smaller weight value.

File: quaterion/train/trainable_model.py [hunk 4]
```diff
@@ -93,6 +95,13 @@ def __init__(self, *args: Any, **kwargs: Any):
         self._model = SimilarityModel(encoders=encoders, head=head)
         self._loss = self.configure_loss()
 
+        self._xbm_config = self.configure_xbm()
+        self._enable_xbm: bool = (
+            self._xbm_config is not None
+        )  # enable xbm only if a XbmConfig is returned
+        if self._enable_xbm:
+            self._xbm_buffer = XbmBuffer(self._xbm_config)
+
```
[the remark this answers is not in the document]
Comment (monatis) [reply quaterion/train/trainable_model.py]: Updated. Also, `embedding_size` is taken directly from `EncoderHead` now instead of a required property in `XbmConfig`.

File: quaterion/train/trainable_model.py [hunk 5]
```diff
@@ -320,6 +345,20 @@ def _common_step(self, batch, batch_idx, stage: TrainStage, **kwargs) -> Tensor:
...
+            and self.trainer.global_step > self._xbm_config.start_iteration
+        ):
+            self._xbm_buffer.queue(embeddings.detach(), targets["groups"].detach())
+
+            memory_embeddings, memory_groups = self._xbm_buffer.get()
+            memory_loss = self.loss(
+                embeddings, targets["groups"], memory_embeddings, memory_groups
+            )
+            loss = loss + self._xbm_config.weight * memory_loss
+
```
[the remark this answers is not in the document]
Comment (monatis) [reply quaterion/train/trainable_model.py]: Moved to a separate method.

Added the comment, and it's also checked in code now.

File: quaterion/train/xbm/xbm_buffer.py [hunk 6]
```diff
@@ -0,0 +1,62 @@
+from typing import Tuple
+
+import torch
+from torch import LongTensor, Tensor
+
+from .xbm_config import XbmConfig, XbmDevice
```
[the remark this answers is not in the document]
Comment (monatis) [reply quaterion/train/xbm/xbm_buffer.py]: Updated

File: quaterion/loss/triplet_loss.py [hunk 7]
```diff
@@ -55,17 +56,30 @@ def get_config_dict(self):
...
+            groups: shape: (batch_size,) - Batch of labels associated with `embeddings`
+            memory_embeddings: shape: (memory_buffer_size, vector_length) - Embeddings stored
+                in a ring buffer. Used only for XBM
+            memory_groups: shape: (memory_buffer_size,) - Groups ids associated with `memory_embeddings`.
+                Used only for XBM
 
         Returns:
             torch.Tensor: Scalar loss value.
         """
+        if memory_embeddings is not None or memory_groups is not None:
```
[the remark this answers is not in the document]
Comment (monatis) [reply quaterion/loss/triplet_loss.py]: Updated.

File: quaterion/loss/triplet_loss.py [hunk 8]
```diff
@@ -124,3 +138,77 @@ def forward(
...
+            memory_embeddings: shape: (memory_buffer_size, vector_length) - Embeddings stored
+                in a ring buffer
+            memory_groups: shape: (memory_buffer_size,) - Groups ids associated with `memory_embeddings`
+
+        Returns:
+            Tensor: zero-size tensor, XBM loss value.
+        """
+        batch_size = embeddings.size(0)
+
+        # Compute similarity matrix
```
[the remark this answers is not in the document]
Comment (monatis) [reply quaterion/loss/triplet_loss.py:164]: Will update it while refactoring.

File: quaterion/train/trainable_model.py [hunk 9]
```diff
@@ -345,19 +345,7 @@ def _common_step(self, batch, batch_idx, stage: TrainStage, **kwargs) -> Tensor:
...
-            and self.trainer.global_step > self._xbm_config.start_iteration
-        ):
-            self._xbm_buffer.queue(embeddings.detach(), targets["groups"].detach())
-
-            memory_embeddings, memory_groups = self._xbm_buffer.get()
-            memory_loss = self.loss(
-                embeddings, targets["groups"], memory_embeddings, memory_groups
-            )
-            loss = loss + self._xbm_config.weight * memory_loss
+        loss = self._maybe_compute_xbm_loss(stage, loss, embeddings, targets)
```
Comment (joein) [inline quaterion/train/trainable_model.py:348]: Nit: Can we make it just `_compute_xbm_loss` and then `loss += self._compute_xbm_loss()`?

File: quaterion/train/trainable_model.py [hunk 10]
```diff
@@ -93,6 +95,10 @@ def __init__(self, *args: Any, **kwargs: Any):
         self._model = SimilarityModel(encoders=encoders, head=head)
         self._loss = self.configure_loss()
 
+        self._xbm_config = self.configure_xbm()
+        if self._xbm_config is not None:
+            self._xbm_buffer = XbmBuffer(XbmConfig, embedding_size=head.output_size)
```
Comment (joein) [inline quaterion/train/trainable_model.py:100]:
```suggestion
            self._xbm_buffer = XbmBuffer(self._xbm_config, embedding_size=head.output_size)
```

File: quaterion/loss/triplet_loss.py [hunk 11]
```diff
@@ -124,3 +138,77 @@ def forward(
...
+                neg_pair_ = sim_mat[i, neg_pair_idx]
+                neg_pair_ = torch.sort(neg_pair_)[0]
+
+                select_pos_pair_idx = torch.nonzero(
+                    pos_pair_ < neg_pair_[-1] + self._margin
+                ).view(-1)
+                pos_pair = pos_pair_[select_pos_pair_idx]
+
+                select_neg_pair_idx = torch.nonzero(
+                    neg_pair_ > max(0.6, pos_pair_[-1]) - self._margin
```
Comment (joein) [inline quaterion/loss/triplet_loss.py:198]: 1) 0.6 shouldn't be a magic constant, consider assigning it to some class variable
2) do we actually want to choose `pos_pair_[-1]` - the best positive, or we want to choose the worst positive, i.e. `pos_pair_[0]` ?

**Comment by monatis** [comment]
It's zero-size tensor (scalar), so regular equal and not-equal operators should work --we use `torch.allclose` and the similar to compare vectors and embeddings to avoid ambiguous meaning of comparing them.

File: tests/eval/losses/test_triplet_loss.py [hunk 12]
```diff
@@ -28,6 +28,16 @@ def test_batch_hard(self):
...
+    def test_xbm(self):
+        loss = TripletLoss(mining="hard")
+        memory_embeddings = torch.rand(size=(100, 3), dtype=torch.float)
+        memory_groups = torch.randint(low=1, high=10, size=(100,), dtype=torch.long)
+        regular_loss = loss.forward(self.embeddings, self.groups)
+        xbm_loss = loss.forward(
+            self.embeddings, self.groups, memory_embeddings, memory_groups
+        )
+
+        assert regular_loss.shape == xbm_loss.shape and regular_loss != xbm_loss
```
[the remark this answers is not in the document]
Comment (joein) [reply tests/eval/losses/test_triplet_loss.py:43]: @monatis I agree, I'm wrong, this statement is ok

File: quaterion/train/xbm/xbm_buffer.py [hunk 13]
```diff
@@ -0,0 +1,62 @@
...
+
+        Args:
+            embeddings: Output embeddings in the batch.
+            targets: Target values in the batch.
+        """
+        batch_size = len(targets)
+
+        if self._pointer + batch_size > self._cfg.buffer_size:
+            self._embeddings[-batch_size:] = embeddings
+            self._targets[-batch_size:] = targets
```
Comment (joein) [inline quaterion/train/xbm/xbm_buffer.py:54]: In the current situation we can replace a part of the (n-1)-th batch:

Let's assume our buffer size is 250 and batch size is 16. 
We put 15 batches (15*16=240 embeddings), and in the next step we want to add  one more batch (16 embeddings). We put them into last 16 cells of the buffer. Hence, we replace 6 embeddings (pos 234-240) from the (n-1)-th batch, but I think it would be better to replace those in the positions 0-6.

Or we can make user to set buffer size proportional to batch size to avoid this problem.

File: quaterion/train/xbm/xbm_buffer.py [hunk 14]
```diff
@@ -0,0 +1,62 @@
...
+from quaterion.train.xbm.xbm_config import XbmConfig, XbmDevice
+
+
+class XbmBuffer:
+    """A buffer implementation to hold recent N embeddings and target values.
+
+    Inspired by https://github.com/msight-tech/research-xbm/blob/master/ret_benchmark/modeling/xbm.py
+
+    Args:
+        config: Config class to configure XBM settings.
```
Comment (joein) [inline quaterion/train/xbm/xbm_buffer.py:15]: please, add `embedding_size` to args, otherwise in docs it seems like there is only one param to manage

File: quaterion/train/xbm/xbm_buffer.py (hunk 13 above)
Comment (monatis) [reply quaterion/train/xbm/xbm_buffer.py:65]: good catch 👍 handled it.

**Review by joein — APPROVED** [review]

**Comment by generall** [comment]
trying to compare xbm_loss and regular loss:

```
torch.random.manual_seed(42)
embeddings = torch.rand(64, 100)
groups = torch.randint(0, 10, (64,))

loss = TripletLoss(margin=1.0, mining="hard")

print(loss(embeddings, groups))
print(loss._compute_xbm_loss(embeddings, groups, embeddings, groups))
```

If I understand the approach right,  loss and `_compute_xbm_loss` should give same result in this case, but in fact I got

```
tensor(1.5443)
tensor(50.5432)
```

Order of magnitude different. I am not sure how the implementation was validated and if this is an expected result, but in this case xbm weight constant is especially tricky to select.
